### ISSUE TITLE
A problem about diff_max setting in your code

### ISSUE THREAD

**Issue by destructive-observer** [riejohnson/cfg-gan-pt#1]
Dear author:
        Thanks for your excellent job. But when i train your code, i find a problem with the value of sum_real minus sum_fake. The gap  will go much bigger like sum_real=50 and sum_fake =-50 when the stage value more than 2000 or 3000,and the generated image will go bad, especially in the lsun_tower dataset..Can you give a explain with the problem? thank you very much.

-------------------log like this-------------------------------------
2021-06-03 17:00:35 EDT: DDG::icfg ... ICFG with cfg_U=1
  t=5: real,62.535987854003906, fake,-58.2081298828125 
  t=10: real,57.689430236816406, fake,-54.70582580566406 
  t=15: real,60.63902282714844, fake,-56.40956115722656 

![lsun_tower64-resnet4-stg07600-16c](https://user-images.githubusercontent.com/10181336/120730029-1e663f80-c513-11eb-8b14-893ad8f29711.jpg)

**Comment by riejohnson** [comment]
Hi.  If you were getting reasonably good generated images before getting to that state, it's likely that your training reached the state of overfitting or overtraining after training too long.  This is explained in C.2.2 of the paper ( http://riejohnson.com/paper/cfggan-tpami.pdf )

" ... if we keep training long enough, the discriminator may eventually overfit as also noted on WGANgp in [11], or it may go to the overtrained state." 

The default number of stages in the code is set to a large number, and I was expecting users to stop training once the quality of generated images plateaus.  But maybe I should change the default to a smaller number now that I think about it.  

If you weren't getting reasonably good generated images before getting to that state, then it's likely that the meta-parameter setting wasn't good.  In that case, it would help to adjust things like cfg_eta, cfg_U, and cfg_T.  

I hope that this is helpful.

**Comment by destructive-observer** [comment]
Thanks for your reply. Is the overfitting or overtraining problem depends on the dataset ?  because when in the lsun_tower , the 3000 stages may reached the line. But in the dining room, only 800 stages get the same problem. But it does not happen in lsun_church svhn cifa10 mnist datasets until 10000 stages (train finish). Is there any theoretic explain with this problem in your cfg model? Or only with practice explain in the experiment?

**Comment by riejohnson** [comment]
Empirical results depend on the combination of dataset, network definition, and the meta-parameter setting.  The default meta-parameter setting of the code is for the datasets tested in the paper, which are MNIST, SVHN, LSUN bedroom + living_room, and LSUN tower + bridge.  "tower" alone or "dining room" was not tested, and so I don't know what meta-parameters are good for them.  The meta-parameter setting for the tested LSUN datasets (br+lr, tower+bridge) are set pretty aggressively to speed up training (taking the risk of instability).  And so what you're seeing could be the results of using meta-parameters that are too aggressive for those datasets.  For example, a larger learning rate and a larger eta are more aggressive, taking big steps in each update.  A smaller T and a smaller N are also aggressive.

**Comment by destructive-observer** [comment]
Thanks for your reply. I will try your advice with more datasets and meta-parameters. Wish you have more excellent job in the future.

**Comment by riejohnson** [comment]
Just one more thing.  

> For example, a larger learning rate and a larger eta are more aggressive, taking big steps in each update. A smaller T and a smaller N are also aggressive.

I might've made it sound complicated by saying this, but in our experiments, simply adjusting eta seemed to be sufficient typically, as long as others were set in a reasonable range.  Thanks for your interest in our work.  I'm now closing this issue ...
